Serialization error if merge() is called without an explicit join

In [this app](https://github.com/RickMohr/treecanopy-ingest/blob/harrisburg-merge/src/main/scala/treecanopy/Main.scala), a "statewide" PA tree canopy raster is merged with a "local" tree canopy raster from Harrisburg.

Note the explicit join [here](https://github.com/RickMohr/treecanopy-ingest/blob/serialization-error/src/main/scala/treecanopy/Main.scala#L77) before the rasters are merged.

[This version](https://github.com/RickMohr/treecanopy-ingest/blob/serialization-error/src/main/scala/treecanopy/Main.scala#L77) is identical except that the rasters are merged without an explicit join. This second version fails with a runtime serialization error:
```
16/06/29 15:21:48 ERROR ApplicationMaster: User class threw exception: org.apache.spark.SparkException: Task not serializable
org.apache.spark.SparkException: Task not serializable
	at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:304)
	at org.apache.spark.util.ClosureCleaner$.org$apache$spark$util$ClosureCleaner$$clean(ClosureCleaner.scala:294)
	at org.apache.spark.util.ClosureCleaner$.clean(ClosureCleaner.scala:122)
	at org.apache.spark.SparkContext.clean(SparkContext.scala:2055)
	at org.apache.spark.rdd.RDD$$anonfun$flatMap$1.apply(RDD.scala:333)
	at org.apache.spark.rdd.RDD$$anonfun$flatMap$1.apply(RDD.scala:332)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:150)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:111)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:316)
	at org.apache.spark.rdd.RDD.flatMap(RDD.scala:332)
	at geotrellis.spark.merge.RDDLayoutMergeMethods.merge(RddLayoutMergeMethods.scala:26)
	at treecanopy.Ingest$$anonfun$main$1.apply(Main.scala:77)
	at treecanopy.Ingest$$anonfun$main$1.apply(Main.scala:74)
	at scala.collection.immutable.Stream.foreach(Stream.scala:547)
	at treecanopy.Ingest$.main(Main.scala:74)
	at treecanopy.Ingest.main(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$2.run(ApplicationMaster.scala:542)
Caused by: java.io.NotSerializableException: geotrellis.spark.package$
Serialization stack:
	- object not serializable (class: geotrellis.spark.package$, value: geotrellis.spark.package$@1c2e0107)
	- field (class: geotrellis.spark.merge.Implicits$withRDDLayoutMergeMethods, name: $outer, type: interface geotrellis.spark.merge.Implicits)
	- object (class geotrellis.spark.merge.Implicits$withRDDLayoutMergeMethods, geotrellis.spark.merge.Implicits$withRDDLayoutMergeMethods@37acb901)
	- field (class: geotrellis.spark.merge.RDDLayoutMergeMethods$$anonfun$1, name: $outer, type: class geotrellis.spark.merge.RDDLayoutMergeMethods)
	- object (class geotrellis.spark.merge.RDDLayoutMergeMethods$$anonfun$1, <function1>)
	at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:40)
	at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:47)
	at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:101)
	at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:301)
	... 20 more
```    
Note that both rasters are small (`raw/treecanopy-2006-2008-pennsylvania-temp` is a small subset of the full statewide raster) so this runs fairly quickly (2 minutes on a 5-node EMR cluster).

--- a/spark/src/main/scala/geotrellis/spark/merge/RDDLayoutMerge.scala
+++ b/spark/src/main/scala/geotrellis/spark/merge/RDDLayoutMerge.scala
@@ -1,0 +1,41 @@
+package geotrellis.spark.merge
+
+import geotrellis.raster._
+import geotrellis.raster.merge._
+import geotrellis.raster.prototype._
+import geotrellis.spark._
+import geotrellis.spark.tiling.LayoutDefinition
+import geotrellis.util._
+
+import org.apache.spark.rdd.RDD
+
+import scala.reflect.ClassTag
+
+object RDDLayoutMerge {
+  /** Merges an RDD with metadata that contains a layout definition into another. */
+  def merge[
+    K: SpatialComponent: ClassTag,
+    V <: CellGrid: ClassTag: ? => TileMergeMethods[V]: ? => TilePrototypeMethods[V],
+    M: (? => LayoutDefinition)
+  ](left: RDD[(K, V)] with Metadata[M], right: RDD[(K, V)] with Metadata[M]) = {
+    val thisLayout: LayoutDefinition = left.metadata
+    val thatLayout: LayoutDefinition = right.metadata
+
+    val cutRdd =
+      right
+        .flatMap { case (k: K, tile: V) =>
+          val extent = thatLayout.mapTransform(k)
+          thisLayout.mapTransform(extent)
+            .coords
+            .map { case (col, row) =>
+              val outKey = k.setComponent(SpatialKey(col, row))
+              val newTile = tile.prototype(thisLayout.tileCols, thisLayout.tileRows)
+              val merged = newTile.merge(thisLayout.mapTransform(outKey), extent, tile)
+              (outKey, merged)
+            }
+        }
+
+
+    left.withContext { rdd => rdd.merge(cutRdd) }
+  }
+}

--- a/spark/src/main/scala/geotrellis/spark/merge/RddLayoutMergeMethods.scala
+++ b/spark/src/main/scala/geotrellis/spark/merge/RddLayoutMergeMethods.scala
@@ -17,26 +17,6 @@ class RDDLayoutMergeMethods[
   M: (? => LayoutDefinition)
 ](val self: RDD[(K, V)] with Metadata[M]) extends MethodExtensions[RDD[(K, V)] with Metadata[M]] {
 
- def merge(other: RDD[(K, V)] with Metadata[M]) = {
-   val thisLayout: LayoutDefinition = self.metadata
-   val thatLayout: LayoutDefinition = other.metadata
-
-   val cutRdd =
-       other
-         .flatMap { case (k: K, tile: V) =>
-           val extent = thatLayout.mapTransform(k)
-           thisLayout.mapTransform(extent)
-             .coords
-             .map { case (col, row) =>
-             val outKey = k.setComponent(SpatialKey(col, row))
-             val newTile = tile.prototype(thisLayout.tileCols, thisLayout.tileRows)
-             val merged = newTile.merge(thisLayout.mapTransform(outKey), extent, tile)
-             (outKey, merged)
-           }
-       }
-
-
-   self.withContext { rdd => rdd.merge(cutRdd) }
- }
-
+ def merge(other: RDD[(K, V)] with Metadata[M]) =
+   RDDLayoutMerge.merge(self, other)
 }

